### ISSUE TITLE
Image editor using Canvas

### ISSUE THREAD

**Issue by riyadhalnur** [keystonejs/keystone#2127]
I've managed to integrate an image editor for images (for single ones not multiple) but I'm having trouble uploading the edited image to S3 or Cloudinary. My current solution is a hack of sorts and not the UX I want to give my users. Any thoughts?

This could be also something to be integrated into core down the line.

**Comment by morenoh149** [comment]
I think that is something firmly outside the scope of keystone.

**Comment by riyadhalnur** [comment]
Maybe core integration, yes but enabling a hackable upload system is not

**Comment by morenoh149** [comment]
you probably want keystone to expose an api endpoint where your clients can perform a put or post request with some image binary data. Try changing the issue name.

**Comment by riyadhalnur** [comment]
Not really an endpoint. I want to be able to overwrite the *file* field for an image so that the image that gets uploaded is not the value of the *file* input field

**Comment by yn5** [comment]
@morenoh149 Why would this be firmly out of Keystone's scope? We would like to be able to crop images in the Admin UI before we upload them to S3 which I think is quite a common feature in CMS's.

@riyadhalnur Can I have a look at your code somewhere?

**Comment by riyadhalnur** [comment]
@yn5 this is what I have so far https://gist.github.com/riyadhalnur/ed7f0b0d6df386680220

**Comment by morenoh149** [comment]
@yn5 when I read the issue it sounded like a cropping tool in the frontend (not adminUI). Glad to hear it does make sense for the adminUI.

**Comment by riyadhalnur** [comment]
@morenoh149 It is for the admin UI - applies to all image fields (S3, Local and so on)

**Comment by yn5** [comment]
@riyadhalnur Could you maybe provide your gist in the form of a Keystone fork?

**Comment by riyadhalnur** [comment]
@yn5 https://github.com/riyadhalnur/keystone/tree/v0.3.x

**Comment by riyadhalnur** [comment]
@yn5 @morenoh149 I've solved the problem of uploading the *base64* data after crop. We need a roadmap to get it finalised for merge into core both for 0.3 and 0.4 branches.

**Comment by yn5** [comment]
@riyadhalnur Haven't had time to have a look into this, hopefully soon I will. Happy to hear you found a fix.

**Comment by riyadhalnur** [comment]
Yeah have it working. We need to come up with a roadmap of sorts and clean up the code.

**Comment by morenoh149** [comment]
just submit a PR when you think it's done. Even better [squash commits](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) before you open a PR.

**Comment by terenzeyuen** [comment]
This sounds ace. Can't wait to see this in 0.3 and 0.4!

**Comment by mxstbr** [comment]
Ref PR: #1747

**Comment by riyadhalnur** [comment]
@mxstbr #1747 is not the reference PR for this one. #1747 is an upgrade to the TinyMCE editor. The editor itself contain plugins for image editing. 

This issue deals with an image editor for image field types.

**Comment by mxstbr** [comment]
Uhh the PR states

> "Added support for the 'imagetools' plugin for TinyMCE. The plugin enables users to edit inline image in the tinyMCE editor"

Is that not the case?

**Comment by riyadhalnur** [comment]
@mxstbr Yeah, the PR states it's for the TinyMCE editor. This issue is specifically for the image field types, which for now are `LocalFile`, `S3File` and 'CloudinaryImage` field types.

**Comment by mxstbr** [comment]
Ah fair enough, my bad, sorry!

**Comment by riyadhalnur** [comment]
@mxstbr No worries :smiley:

**Comment by mxstbr** [comment]
@riyadhalnur are you ready to submit a PR for this against `master`? Would love to land this in 0.4!

**Comment by riyadhalnur** [comment]
@mxstbr Yeah I need to clean up the code and add a few tests. I hope to be done soon.

**Comment by terenzeyuen** [comment]
Can't wait.

**Comment by mxstbr** [comment]
Awesome!!

**Comment by arlejeun** [comment]
I'm using keystone 0.4 and trying to display an image thumbnail in admin UI with Local File type. I've seen a lot of description in github and I believe it is related to 0.3 but I would like to know if it is possible with 0.4. Any hints are welcome.

**Comment by riyadhalnur** [comment]
Will wait till the field types are re-written. Correct me if I'm wrong @mxstbr, @morenoh149 and @JedWatson. I can't seem to find the related issue.

**Comment by kimasendorf** [comment]
@riyadhalnur Awesome, could you point me to the file you have edited/modified to add the image editor?

**Comment by ghuroo** [comment]
I'm also interested in the solution :)

**Comment by tomtom87** [comment]
@riyadhalnur Hi, did you finish your patch? Because I have just started coding something for my project, would like to start making some plugins for Keystone to share this kind of development with the community.

**Comment by tomtom87** [comment]
@arlejeun I had to hack something together to get it working with localfile, adding image preview would be no issue. Do you still need this?

**Comment by ptcampbell** [comment]
Same here 👋
